felix/calc/vxlan_resolver.go

```diff
@@ -171,119 +189,205 @@ func (c *VXLANResolver) OnHostConfigUpdate(update api.Update) (_ bool) {
+			delete(c.nodeNameToSentVTEP, node)
+			c.callbacks.OnVTEPRemove(node)
+			return
+		}
+		logCtx.Info("Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet")
+		return
 	}
 
-	logCxt.Debug("Sending VTEP to dataplane")
 	vtep := &proto.VXLANTunnelEndpointUpdate{
-		Node:           node,
-		ParentDeviceIp: parentDeviceIP,
-		Mac:            c.vtepMACForHost(node),
-		Ipv4Addr:       tunlAddr,
+		Node: node,
+	}
+	if hasV4Info {
+		vtep.ParentDeviceIpv4 = c.nodeNameToIPv4Addr[node]
+		vtep.MacV4 = c.vtepMACForHost(node, 4)
+		vtep.Ipv4Addr = c.nodeNameToVXLANTunnelAddr[node]
+	}
+	if hasV6Info {
+		vtep.ParentDeviceIpv6 = c.nodeNameToIPv6Addr[node]
+		vtep.MacV6 = c.vtepMACForHost(node, 6)
+		vtep.Ipv6Addr = c.nodeNameToVXLANTunnelAddrV6[node]
+	}
+
+	if hasSentVTEP {
+		if c.vtepEqual(oldVTEP, vtep) {
+			logCtx.Info("VTEP information has not changed, skipping duplicate update")
+			return
+		}
+		// Skip removing node from c.nodeNameToSentVTEP because it will be updated below
+		// Send a remove for the old VTEP information
+		c.callbacks.OnVTEPRemove(node)
 	}
+
+	logCtx.Info("Sending VTEP update")
+	c.nodeNameToSentVTEP[node] = vtep
 	c.callbacks.OnVTEPUpdate(vtep)
-	return true
 }
 
-func (c *VXLANResolver) sendVTEPRemove(node string) {
-	logrus.WithField("node", node).Debug("Withdrawing VTEP from dataplane")
-	c.callbacks.OnVTEPRemove(node)
+// vtepEqual compares if 2 proto.VXLANTunnelEndpointUpdate messages contain the same information
+func (c *VXLANResolver) vtepEqual(vtep1, vtep2 *proto.VXLANTunnelEndpointUpdate) bool {
+	switch {
+	case vtep1.Node != vtep2.Node:
+		return false
+	case vtep1.MacV4 != vtep2.MacV4:
+		return false
+	case vtep1.Ipv4Addr != vtep2.Ipv4Addr:
+		return false
+	case vtep1.ParentDeviceIpv4 != vtep2.ParentDeviceIpv4:
+		return false
+	case vtep1.MacV6 != vtep2.MacV6:
+		return false
+	case vtep1.Ipv6Addr != vtep2.Ipv6Addr:
+		return false
+	case vtep1.ParentDeviceIpv6 != vtep2.ParentDeviceIpv6:
+		return false
+	}
+
+	return true
 }
 
 // vtepMACForHost checks if there is new MAC present in host config.
 // If new MAC is present in host config, then vtepMACForHost returns the MAC present in  host config else
 // vtepMACForHost calculates a deterministic MAC address based on the provided host.
 // The returned address matches the address assigned to the VXLAN device on that node.
-func (c *VXLANResolver) vtepMACForHost(nodename string) string {
-	mac := c.nodeNameToVXLANMac[nodename]
+func (c *VXLANResolver) vtepMACForHost(nodename string, ipVersion int) string {
+	logCtx := logrus.WithFields(logrus.Fields{"node": nodename, "ipVersion": ipVersion})
+	var mac string
+
+	switch ipVersion {
+	case 4:
+		mac = c.nodeNameToVXLANMac[nodename]
+		if mac != "" {
+			return mac
+		}
+	case 6:
+		mac = c.nodeNameToVXLANMacV6[nodename]
+		if mac != "" {
+			return mac
+		}
 
-	if mac != "" {
-		return mac
+		nodename += "-v6"
```

Comment:
I think this should be fine but just wanted to see if you checked that this suffix makes a difference for the for SHA-1 hash since it seems we only use the first 5 bytes?

Reply:
In the tests it does generate different MACs, we generate the hash from the whole string and then just use the first 5 bytes, right? I suppose SHA-1 is sufficiently robust that we can trust its "avalanche effect" to change the hash significantly...